zilliqa/src/pool.rs

```diff
@@ -67,75 +116,105 @@ impl TransactionPool {
+                // or `pop_transaction`), but a stale reference would still exist in the heap.
+                //
                 // We loop until we find a transaction that hasn't been made invalid.
                 continue;
             };
-            self.hash_to_index.remove(&transaction.hash);
 
-            if let Some(next_txn) = self.transactions.get(&(from_addr, nonce + 1)) {
-                self.ready.push(next_txn.into());
+            // If we've popped a nonced transaction, that may have made a subsequent one valid
+            if let TxIndex::Nonced(address, nonce) = tx_index {
+                if let Some(next_txn) = self.transactions.get(&TxIndex::Nonced(address, nonce + 1))
+                {
+                    self.ready.push(next_txn.into());
+                }
             }
 
             return Some(transaction);
         }
     }
 
     pub fn insert_transaction(&mut self, txn: VerifiedTransaction, account_nonce: u64) -> bool {
-        if txn.tx.nonce() < account_nonce {
+        if txn.tx.nonce().is_some_and(|n| n < account_nonce) {
             // This transaction is permanently invalid, so there is nothing to do.
             return false;
         }
 
-        if let Some(existing_txn) = self.transactions.get(&(txn.signer, txn.tx.nonce())) {
-            // There is already a transaction in the mempool with the same signer and nonce. Only proceed if this one
-            // is better. Note that if they are equally good, we prioritise the existing transaction to avoid the need
-            // to broadcast a new transaction to the network.
-            if ReadyItem::from(existing_txn) >= ReadyItem::from(&txn) {
+        if let Some(tx_nonce) = txn.tx.nonce() {
+            if let Some(existing_txn) = self
+                .transactions
+                .get(&TxIndex::Nonced(txn.signer, tx_nonce))
+            {
+                // There is already a transaction in the mempool with the same signer
+                // and nonce. Only proceed if this one is better. Note that if they are
+                // equally good, we prioritise the existing transaction to avoid the need
+                // to broadcast a new transaction to the network.
+                if ReadyItem::from(existing_txn) >= ReadyItem::from(&txn) {
+                    return false;
+                } else {
+                    // Remove the existing transaction from `hash_to_index` if we're about to replace it.
+                    self.hash_to_index.remove(&existing_txn.hash);
+                }
+            }
+        } else {
+            // now we've confirmed it's nonceless, so ensure its index isn't duplicate
+            if let Some(existing_nonceless_txn) = self.transactions.get(&txn.mempool_index()) {
+                warn!(tx = ?existing_nonceless_txn, "Duplicate-indexed nonceless transactions encountered, this shouldn't happen. Ignoring the new one.");
```

Comment:
How do we ensure nonces aren't duplicated later on, once transactions have left the mempool and been committed to the chain?

Reply:
We don't - in theory we assume that the local sister node sending us the local `IntershardCall` notifications should never double-send the same one. This check is more of a sanity check for transient local errors - `"this shouldn't happen"` is not simply wishful, it's meant to be literal.
Otherwise the bridge nonce is not really meant to provide replay protection. The source shard is responsible for that.